contents/projectManagement/testDrivenDevelopment.md

```diff
@@ -0,0 +1,207 @@
+Mock objects are simulated objects whose sole purpose is to act as a black box that mimics the behaviour of real-world objects during testing. They are especially useful in writing unit tests, which are meant to test the functionality of a portion of code assuming that the rest of the code behaves as it should. 
+</modal>
+
+### Benefit 2: Ensures that Product Requirements are Understood
+
+Test-driven development requires the developers to have a good understanding of the product or feature requirements before any sort of development can begin. As mentioned earlier, the first step in development is writing tests. Coming up with tests without a clear picture of the requirements or specifications  can be dangerous as: 
+
+1. you could end up testing something that doesn't need to be tested, or 
+1. you could end up overlooking some requirements that _must_ be tested, thus giving you a false sense of accomplishment when you write production code to pass the test
+
+TDD thus prevents the team from rushing into development and pushes them to prioritize the requirements-gathering step. 
+
+
+
+### Benefit 3: Increases Productivity
+
+Hakan Erdogmus, Professor of Software Engineering at Carnegie Mellon University, found that programmers who wrote more tests tended to be more productive^[Erdogmus, 2005. ["On the Effectiveness of Test-first Approach to Programming"](https://web.archive.org/web/20141222180731/http://nparc.cisti-icist.nrc-cnrc.gc.ca/npsi/ctrl?action=shwart&index=an&req=5763742&lang=en)], and since employing TDD meant writing more tests, there is a correlation between productivity and TDD.
+
+Employing TDD also lays out the design of a program. It forces programmers to think about coming up with failing tests and working their way up to writing code that passes the test.
+
+Furthermore, since tests are written first, there is close to 100% coverage. This ensures that production code is covered by at least one test. This can boost individual and team morale, confidence in the code, and productivity as a consequence.
+
+
+## Caveats and Limitations
+
+Even with some of the attractive benefits of adopting TDD, there is no one-size-fits-all solution for project design management. Some of the limitations of TDD are: 
+
+1. **Strict emphasis on communication of requirements and specifications**
+
+    Product/Feature requirements are crucial to teams that employ a TDD-style approach to development. As mentioned earlier, the very first step of the development cycle is to write acceptance tests. If the requirements are not communicated properly, teams will end up constantly having to change their test suites. In extreme scenarios, a re-write may be necessary.
+
+    As a result, a lot of man hours can go into making sure the requirements are in order -  consequently, this could result in a slow start to development.
+
+2. **Tough to incorporate TDD into <trigger trigger="click" for="modal:index-legacy">legacy projects [^6]</trigger>**
+
```

Comment:
A diagram illustrating a problematic architecture of a legacy project here and how it leads harder-to-write tests could be nice here